### PR TITLE
Experimental Nix support

### PR DESCRIPTION
**Pull request by godexsoft**
This PR adds support for building and/or developing `libfn` using Nix.
Some details on how to get it to do its thing are provided in NIX.md.
Hopefully this does not break anything else. At least CI seems happy now and I was able to still build stuff from my devcontainer setup.

### CODE REVIEW

File: catch2_3.nix [hunk 1]
```diff
@@ -0,0 +1,38 @@
+{ lib
+, stdenv
+, fetchFromGitHub
+, cmake
+, python3
+}:
+
+stdenv.mkDerivation rec {
+  pname = "catch2_local";
+  version = "3.5.2";
```
Comment (Bronek) [inline catch2_3.nix:10]: can we use the same version as in `Catch2.cmake` i.e. 3.5.4 ?

**Review by Bronek — COMMENTED** [review]
I know nothing about building packages with `nix` so feel free to drop this suggestion, but would it be possible to move `.nix` files into a new `nix` directory ?

**Comment by godexsoft** [comment]
> I know nothing about building packages with `nix` so feel free to drop this suggestion, but would it be possible to move `.nix` files into a new `nix` directory ?

I'd love to. I tried to do so a few weeks back but being a nix noob myself i failed miserably so had to revert it. I'll try again as i'd also prefer the root of the project to remain as uncluttered as possible.

**Comment by godexsoft** [comment]
> > I know nothing about building packages with `nix` so feel free to drop this suggestion, but would it be possible to move `.nix` files into a new `nix` directory ?
> 
> I'd love to. I tried to do so a few weeks back but being a nix noob myself i failed miserably so had to revert it. I'll try again as i'd also prefer the root of the project to remain as uncluttered as possible.

So i was able to move most files out of the way and into `nix`. But the main `flake` and its lockfile apparently need to be at the root of the repo if we want to be able to avoid having to do something like `nix build github:libfn/functional?dir=nix#clang`. For the build environment it would probably look like `nix develop ./nix#clang` or something but i have not tried.

Let me know if you are ok with the flake in root or if you would like me to do more research on this.

**Comment by godexsoft** [comment]
btw., we could totally add all the remaining tooling to the build environment. Stuff like `clangd`, `clang-tidy`, `clang-format` and whatever else we need to develop the library comfortably. But this could also be done in a separate branch later on.

File: include/CMakeLists.txt [hunk 2]
```diff
@@ -38,3 +38,10 @@ set(INCLUDE_SOURCE_FILES
 
 add_library(${PROJECT_NAME} INTERFACE ${INCLUDE_SOURCE_FILES})
 target_include_directories(${PROJECT_NAME} INTERFACE ${CMAKE_CURRENT_SOURCE_DIR})
+
+set(INSTALL_DESTINATION ${CMAKE_INSTALL_PREFIX}/libfn)
```
Comment (Bronek) [inline include/CMakeLists.txt:42]: I think that should be without the `/libfn` suffix, and instead ...

File: include/CMakeLists.txt [hunk 3]
```diff
@@ -38,3 +38,10 @@ set(INCLUDE_SOURCE_FILES
...
+
+install(FILES ${INCLUDE_SOURCE_FILES} 
+        DESTINATION ${INSTALL_DESTINATION}/functional)
```
Comment (Bronek) [inline include/CMakeLists.txt:45]: ... this destination should be `${INSTALL_DESTINATION}/include/functional` and also ...

File: include/CMakeLists.txt [hunk 4]
```diff
@@ -38,3 +38,10 @@ set(INCLUDE_SOURCE_FILES
...
+install(FILES ${CMAKE_SOURCE_DIR}/README.md ${CMAKE_SOURCE_DIR}/LICENSE.md 
+        DESTINATION ${INSTALL_DESTINATION})
```
Comment (Bronek) [inline include/CMakeLists.txt:47]: ... this destination should be `${CMAKE_INSTALL_PREFIX}/share/doc/functional`

**Comment by Bronek** [comment]
> > > I know nothing about building packages with `nix` so feel free to drop this suggestion, but would it be possible to move `.nix` files into a new `nix` directory ?
> > 
> > 
> > I'd love to. I tried to do so a few weeks back but being a nix noob myself i failed miserably so had to revert it. I'll try again as i'd also prefer the root of the project to remain as uncluttered as possible.
> 
> So i was able to move most files out of the way and into `nix`. But the main `flake` and its lockfile apparently need to be at the root of the repo if we want to be able to avoid having to do something like `nix build github:libfn/functional?dir=nix#clang`. For the build environment it would probably look like `nix develop ./nix#clang` or something but i have not tried.
> 
> Let me know if you are ok with the flake in root or if you would like me to do more research on this.

I'm OK with that.

File: include/CMakeLists.txt (hunk 3 above)
Comment (Bronek) [reply include/CMakeLists.txt:45]: Actually this should be a relative path. CMake makes it implicitly relative to either of `$CMAKE_INSTALL_PREFIX` or `--prefix` parameter of the `install` target

File: include/CMakeLists.txt (hunk 4 above)
Comment (Bronek) [reply include/CMakeLists.txt:47]: this should be also relative path, like so `DESTINATION share/doc/functional`

**Comment by Bronek** [comment]
@godexsoft as you probably noticed, I change `install` targets

**Comment by Bronek** [comment]
@godexsoft there's a problem:

```
bronekk@aleph ~ % nix run --extra-experimental-features nix-command --extra-experimental-features flakes github:libfn/functional/feature/nix-build
error: unable to execute '/nix/store/x2kbhqgfnbdnr7s7fmk02yncnfv45m9j-libfn/bin/libfn': No such file or directory
```

**Comment by godexsoft** [comment]
> @godexsoft there's a problem:
> 
> ```
> bronekk@aleph ~ % nix run --extra-experimental-features nix-command --extra-experimental-features flakes github:libfn/functional/feature/nix-build
> error: unable to execute '/nix/store/x2kbhqgfnbdnr7s7fmk02yncnfv45m9j-libfn/bin/libfn': No such file or directory
> ```

I think it's because there is nothing to run? We are not installing any executables as far as i know. 
When i use `nix build github:libfn/functional/feature/nix-build -L` it appears to do what we want. I can see it's now installing to the paths you setup. Once you do a `nix build` from the github url it places all outputs into the `result` folder on your CWD.  You can try to replace `run` with `build` in your command and maybe add `-L` that will produce all the output to the console instead of just one line.

**Pull request merged**